### PR TITLE
Add score function inference

### PR DESCRIPTION
**Pull request by meissnereric**
*Issue #18 *

*Description of changes:*
This version has the basic score function inference algorithm with Rao-Blackwellization implemented. 

This PR also fixes a bug in the MultivariateNormal class causing it to only ever draw sample once. 

Tests have been added for both changes.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: mxfusion/inference/score_function.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        q_z_lambda = self.posterior.compute_log_prob(
+            F=F, targets=knowns, constants=constants)
+
+        p_x_z = self.model.compute_log_prob(
+            F=F, targets=knowns, constants=constants)
+
+        difference_nograd = F.stop_gradient(p_x_z - q_z_lambda)
+        gradient_lambda = F.mean(q_z_lambda * difference_nograd, axis=0)
+
+        gradient_theta = F.mean(p_x_z - F.stop_gradient(q_z_lambda), axis=0)
```
Comment (zhenwendai) [inline mxfusion/inference/score_function.py]: It should be equivalent to the follows:
```python
gradient_theta = F.mean(p_x_z, axis=0)
```

File: mxfusion/inference/score_function.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        # Robbins-Monro sequence??
+        gradient_log_L = gradient_lambda + gradient_theta
+
+        return -gradient_theta, -gradient_log_L
+
+    def _extract_descendant_blanket_params(self, graph, node):
+        """
+        Returns a set of the markov blankets of all of the descendents of the node in the graph, mapped to their parameter form.
+        """
+        if node.graph != graph.components_graph:
```
Comment (zhenwendai) [inline mxfusion/inference/score_function.py]: Is this the expected way in Python to check two reference points to the same data structure?

File: mxfusion/inference/score_function.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+        gradient_theta = F.mean(p_x_z - F.stop_gradient(q_z_lambda), axis=0)
+        posterior_rvs = [v for v in self.posterior.variables.values() if v.type is VariableType.RANDVAR]
+        f_list = []
+        grad_list = []
+        for i, v in enumerate(posterior_rvs):
+            model_v = self.model[v]
+
+            q_i_varset = self._extract_descendant_blanket_params(self.posterior, v)
+            q_i_params = {key:val for key,val in knowns.items() if key in q_i_varset}
```
Comment (zhenwendai) [inline mxfusion/inference/score_function.py]: Why not do the following?
```python
q_i_params = {key: knowns[key] for key in q_i_varset}
```

File: testing/inference/score_function_test.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+import mxfusion as mf
+import mxnet as mx
+import numpy as np
+import mxnet.gluon.nn as nn
+from mxfusion.components.variables.var_trans import PositiveTransformation
+from mxfusion.components.functions import MXFusionGluonFunction
+from mxfusion.util.testutils import make_basic_model
+
+
+class ScoreFunctionTests(unittest.TestCase):
```
Comment (zhenwendai) [inline testing/inference/score_function_test.py]: We need a test case for the correctness of the inference method, maybe compare with the gradient from the ones from reparameterization tricks.

File: mxfusion/inference/score_function.py (hunk 2 above)
Comment (meissnereric) [reply mxfusion/inference/score_function.py]: As long as the __eq__ operator isn't overridden by the class then yeah, it is the same as call id(object) == id(object2). So this is fine I think. Could do node.graph is graph.components_graph if we wanted to always use id() comparison.

File: mxfusion/inference/score_function.py (hunk 1 above)
Comment (meissnereric) [reply mxfusion/inference/score_function.py]: Changed.

File: mxfusion/inference/score_function.py (hunk 3 above)
Comment (meissnereric) [reply mxfusion/inference/score_function.py]: This isn't the same, since not all variables in q_i_varset are in knowns (i.e. the unknown ones.)

**Pull request merged**